File: OC/TableViewBUGOC/TableViewBUGOC/ViewController.m
//
//  ViewController.m
//  TableViewBUGOC
//
//  Created by study on 17/3/31.
//  Copyright © 2017年 study. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property(nonatomic , strong)UITableView *tableView;
@end

@implementation ViewController


- (IBAction)myClick:(UIButton *)sender {
    
    NSIndexPath *index1 = [NSIndexPath indexPathForRow:2 inSection:0];
    NSIndexPath *index2 = [NSIndexPath indexPathForRow:3 inSection:0];
    
    [self.tableView reloadRowsAtIndexPaths:@[index1 , index2] withRowAnimation:(UITableViewRowAnimationNone)];
    
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView = [[UITableView alloc]initWithFrame:self.view.frame style:(UITableViewStylePlain)];
    [self.view insertSubview:self.tableView atIndex:0];
    self.tableView.delegate =self;
    self.tableView.dataSource = self;
    // Do any additional setup after loading the view, typically from a nib.
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return 20;
}

-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    
    NSString *text = [NSString stringWithFormat:@"section =%li row = %li" , (long)indexPath.section , (long)indexPath.row];
    NSLog(@"%@" , text);
    
    if ((indexPath.row & 1) == 0) {
        return 45;
    }else{
        return 100;
    }
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    NSString *ids = @"cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:ids];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:(UITableViewCellStyleDefault) reuseIdentifier:ids];
    }
    NSString *text = [NSString stringWithFormat:@"section =%li row = %li" , (long)indexPath.section , (long)indexPath.row];
    
    cell.textLabel.text = text;
    return cell;
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: Swift/dddddd/ViewController.swift
//
//  ViewController.swift
//  dddddd
//
//  Created by study on 17/3/30.
//  Copyright © 2017年 study. All rights reserved.
//

import UIKit

class ViewController: UIViewController , UITableViewDelegate , UITableViewDataSource {

    @IBAction func myClick(_ sender: Any) {
        
        let index1 = IndexPath(row: 2, section: 0);
        let index2 = IndexPath(row: 3, section: 0);
        
        tableView.reloadRows(at: [index1 , index2 ], with: .none)
    }
    
    var tableView:UITableView!
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView = UITableView(frame: self.view.frame, style: UITableViewStyle.plain)
        
        self.view.insertSubview(tableView, at: 0)
        
        tableView.delegate = self
        tableView.dataSource = self
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 20
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        
        print("section \(indexPath.section) row=\(indexPath.row)")
       
        if indexPath.row & 1 == 0 {
            return 45
        }else{
            return 100
        }
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let id = "cell"
        var cell = tableView.dequeueReusableCell(withIdentifier: id)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: id)
        }
        cell?.textLabel?.text = "section \(indexPath.section) row=\(indexPath.row)"
        
        return cell!
    }
    
}

